src/Mvc/Mvc.TagHelpers/src/AttributeMatcher.cs

```diff
@@ -47,13 +46,11 @@ internal static class AttributeMatcher
             result = default;
 
             // Perf: Avoid allocating enumerator
-            var modeInfosCount = modeInfos.Count;
             var allAttributes = context.AllAttributes;
             // Read interface .Count once rather than per iteration
             var allAttributesCount = allAttributes.Count;
-            for (var i = 0; i < modeInfosCount; i++)
+            foreach (var modeInfo in modeInfos.AsSpan())
```

Comment:
You don't need the `AsSpan`, foreaching over an array is pretty efficient.

Reply:
See https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgMwAEu5AwuQN7HlOUVYAs5AsgBTUDaAXXIATCDAgBKeo2ayAZgHsEcCAGMAFtwBuEBOVXkAlgDsRYyTNlMGRK3f0A6ACIKAkjG4SA3JasBfXwDbZl8ySg5OHF5BM3EpG3tyRWU1TR09AxNYiAcAQSgAZQAHCGNPCV9ZBMTmVWc3D29K5iDZVpDgpjD2chd3T19qpiC/IA

More efficient actually 😄

Reply:
Interestingly, foreaching over and array and an explicit for loop over an array are equivalent.

Reply:
> Interestingly, foreaching over and array and an explicit for loop over an array are equivalent.

C# compiler lowers the foreach over array to a for-loop (I believe this is done since the beginning).

Reply:
Yep, you can see it in the C# -> C# translation.

Reply:
It does the same for Span\<T\> too

Reply:
So:
- T[]  - foreach or for
- Span\<T\> - foreach and for
- List<T> - CollectionsMarshal.AsSpan

Reply:
In the T[] case for really (safe) high-perf `for` has the advantage that one can index with `uint`s, so no `movsxd` (sign extensions) needs to be emitted. Unfortunately current JIT (at least the one from sharplab) got "de-optimized", this is tracked in https://github.com/dotnet/runtime/issues/35618 now.